### PR TITLE
coalescing start transaction calls into a single timelock rpc

### PR DESCRIPTION
**Pull request by gokcanozakdag**
**Goals (and why)**:
To reduce timelock load created by start transaction calls, as we are expecting an increase in start transaction calls after moving to thorough sweep (as read only transaction will be required to grab immutable ts locks)

**Implementation Description (bullets)**:
Using `AutoBatcher` to coalesce calls into a single batched call. Using `LockTokenShare` to enable reference counting on immutable ts lock token.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Testing `CoalescingTransactionService` for both single and multi threaded calls.

**Concerns (what feedback would you like?)**:
I will try to refactor it a bit more to make it easier to follow.
We first need to release Timelock with Atlas 0.126.0, before merging this pr.

**Where should we start reviewing?**:
`CoalescingTransactionService`

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 1]
```diff
@@ -32,10 +36,12 @@
 public final class RemoteTimelockServiceAdapter implements TimelockService {
```
Comment (felixdesouza) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java]: I've found the logic here pretty hard to follow to be honest, haven't really gotten a good idea of what it should look like just yet, but yeah.

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 2]
```diff
@@ -32,10 +36,12 @@
...
     private final LockLeaseService lockLeaseService;
     private final TimelockRpcClient timelockRpcClient;
+    private final CoalescingTransactionService transactionService;
 
     private RemoteTimelockServiceAdapter(TimelockRpcClient timelockRpcClient) {
         this.timelockRpcClient = timelockRpcClient;
         this.lockLeaseService = LockLeaseService.create(timelockRpcClient);
+        this.transactionService = CoalescingTransactionService.create(lockLeaseService);
```
Comment (felixdesouza) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java]: `startTransactionService`

File: lock-api/src/test/java/com/palantir/lock/client/CoalescingTransactionServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    private static PartitionedTimestamps getPartitionedTimestamps(long startTs, int count) {
+        return ImmutablePartitionedTimestamps.builder()
+                .start(startTs)
+                .count(count)
+                .interval(NUM_PARTITIONS)
+                .build();
+    }
+
+    private static final class BlockingBatchedResponse implements Answer<StartTransactionResponseV4> {
```
Comment (felixdesouza) [inline lock-api/src/test/java/com/palantir/lock/client/CoalescingTransactionServiceTest.java]: We have tests for autobatcher, I'm not sure it's useful to prove that it works everywhere we use it imo.

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.common.base.Throwables;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartTransactionResponseV4;
+import com.palantir.lock.v2.TimestampAndPartition;
+
+final class CoalescingTransactionService {
```
Comment (felixdesouza) [inline lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: The more I think about, the more I think that this class needn't exist. We should just autobatch by default inside `LockLeaseService`. The contract is the same, we don't need to test that it will work, since we have tests for autobatcher. Akin to caching.

Once you've moved out all the different misplaced pieces, it actually doesn't do much besides tunnel calls into autobatcher. The same way we wouldn't test a cache (unless that cache is doing different things such as in `CachingPartitionAllocator`), I don't think it's necessary to test autobatcher.

Comment (felixdesouza) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: This is a slightly wild view, so can do this last, but I think overall it does make things cleaner.

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    private final DisruptorAutobatcher<Void, StartIdentifiedAtlasDbTransactionResponse> autobatcher;
+
+    private CoalescingTransactionService(
+            DisruptorAutobatcher<Void, StartIdentifiedAtlasDbTransactionResponse> autobatcher) {
```
[the remark this answers is not in the document]
Comment (j-baker) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: dunno about you, but generally I get the heebyjeebies when I see a constructor which actually does work. I end up struggling to think about initialization races, data sharing, etc etc, and then the second constructor must take care to make this work.

Much rather we just inject static state like this PR does currently, right?

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.common.base.Throwables;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartTransactionResponseV4;
+import com.palantir.lock.v2.TimestampAndPartition;
+
+final class CoalescingTransactionService {
+    private final DisruptorAutobatcher<Void, StartIdentifiedAtlasDbTransactionResponse> autobatcher;
```
Comment (j-baker) [inline lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: note: these autobatchers are closeable, and you have to close them. It won't work to just create them, you'll end up with a memory leak.

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        List<StartIdentifiedAtlasDbTransactionResponse> result = new ArrayList<>();
+        while (result.size() < numberOfTransactions) {
+            result.addAll(split(lockLeaseService.startTransactions(numberOfTransactions - result.size())));
+        }
+        return result.subList(0, numberOfTransactions);
+    }
+
+    @VisibleForTesting
+    static List<StartIdentifiedAtlasDbTransactionResponse> split(StartTransactionResponseV4 batchedResponse) {
+        LockImmutableTimestampResponse immutableTsAndLock = batchedResponse.immutableTimestamp();
```
[the remark this answers is not in the document]
Comment (gokcanozakdag) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: hmm, actually moving split logic to StartTransactionReponse requires making LockTokenShare public rather than package private, which I really don't want to.

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+
+    private static List<StartIdentifiedAtlasDbTransactionResponse> startTransactions(
+            LockLeaseService lockLeaseService, int numberOfTransactions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> result = new ArrayList<>();
+        while (result.size() < numberOfTransactions) {
+            result.addAll(split(lockLeaseService.startTransactions(numberOfTransactions - result.size())));
+        }
+        return result.subList(0, numberOfTransactions);
```
[the remark this answers is not in the document]
Comment (gokcanozakdag) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: not really, too much of paranoia

File: lock-api/src/main/java/com/palantir/lock/client/LockLeaseService.java [hunk 9]
```diff
@@ -35,10 +35,12 @@
 import com.palantir.lock.v2.StartAtlasDbTransactionResponseV3;
 import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionRequest;
 import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartTransactionRequestV4;
+import com.palantir.lock.v2.StartTransactionResponseV4;
 import com.palantir.lock.v2.TimelockRpcClient;
 import com.palantir.logsafe.Preconditions;
 
-final class LockLeaseService {
+class LockLeaseService {
```
[the remark this answers is not in the document]
Comment (gokcanozakdag) [reply lock-api/src/main/java/com/palantir/lock/client/LockLeaseService.java]: mockito was complaining

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java (hunk 6 above)
Comment (gokcanozakdag) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: will add that after merging #3841

File: lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java (hunk 4 above)
Comment (gokcanozakdag) [reply lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java]: I have concerns on adding the coalescing logic to lockLeaseService. I believe it would make it pretty hard to follow, but functionalities are complex enough to have its own place. Of course, this layered approach is not the only way of decomposing the functionality, but in any case I believe we need to separate them in some way. Tried something different here: #3844
